src/webview/html/messageAsHtml.js

```diff
@@ -72,13 +77,87 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const widgetSubmessages = message.submessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      ...submessage,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const poll_widgets = widgetSubmessages.filter(widget => widget.content.widget_type === 'poll');
+  // "events" are submessages other than the main widget submessage - for
+  // instance, `vote`, `new_option`, and similar submessages. These are not
+  // documented, look at the PollData code if you're interested in what the
+  // possibilities are.
+  const events = widgetSubmessages.filter(widget => 'type' in widget.content);
+
+  if (poll_widgets.length === 0) {
+    return template`
 $!${message.content}
 <div class="special-message"
  ><p>Interactive message</p
  ><p>To use, open on web or desktop</p
 ></div>
 `;
+  }
+  const poll_widget = poll_widgets[0];
+
+  if (!poll_widget.content || !poll_widget.content.extra_data) {
+    // We don't expect this to happen in general, but there are some malformed
+    // messages lying around that will trigger this.
+    return template`$!${message.content}`;
+  }
+
+  const poll_data = new PollData({
+    current_user_id: ownUserId,
+    is_my_poll: message.sender_id === ownUserId,
+    question: poll_widget.content.extra_data.question,
+    options: poll_widget.content.extra_data.options,
+    comma_separated_names: () => {},
```

Comment:
This should get a TODO comment -- it's a stub we'll need to fill in to get the feature fully working.

Also it should return a string, probably "".

Reply:
Done. FWIW, I expect to change this to operate based on user IDs when we actually build this feature.